### PR TITLE
Fetch radio schedules data on server

### LINKED ISSUE #5537: Radio Schedules: Server Render component
**Is your feature request related to a problem? Please describe.**
We will have a Radio Schedules container, that fetches radio schedules data from the client, and then renders the Radio Schedules component. This could cause content to jump in the viewport, so we should server render this component first.

We have implemented a solution for this as part of https://github.com/bbc/simorgh/issues/4472 - follow the same pattern to implement this for radio schedules.

There is a difference in that we will need to fetch the initial data as extra data for another page type, rather than a whole new page as we have with most read. So we should consider how we can write this data fetching code in a way that can be included in getInitialData for existing page types, and reused across the page types that will need this data.

**Describe the solution you'd like**
Reuse the data fetching and processing that is used on the client on the server, and pass this data into the component on the server.

**Describe alternatives you've considered**

**Testing notes**
[Tester to complete]

- Unit tests
- Uncomment the integration test that has been written for radio schedules, as it can now be run against the server rendered component.

- [x] This feature is expected to need manual testing.

**Additional context**

### PR DESCRIPTION
**Pull request by tochwill**
Resolves #5537

In order to render the radio schedule for a service on the server, we need to fetch the data upfront, before we render the react application. Here, we fetch that data in parallel with the page data required to render a front page with radio schedules.

**Overall change:**
This PR adds a call on the server to fetch radio schedule data for front pages that require it. If we get the data back from the API call, we merge it into the page data that is passed into the application. This allows us to pass this data into the radio schedules component in order to be rendered on the server. The component is still able to fetch the data it needs on the client if it has not been passed data.

**Code changes:**

- Adds a 'withRadioSchedule' function, that wraps the data fetching promise to fetch page data, and merges in the radio schedule data. The getInitialData function first checks to see if a service requires radio schedules on the front page, and if it does, it uses this function. Otherwise it fetches page data as previously.

- Adds an initialData prop to the radioSchedule container, which allows for data fetched on the server to be passed in and rendered.

- Pulls the radio schedule data out of pageData in the front page container, and passes it down to the radio schedule container.

- Updates tests and data processing utility functions.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] I have added labels to this PR for the relevant pod(s) affected by these changes
- [x] I have assigned this PR to the Simorgh project

**Testing:**

- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`)
-  [x] This PR requires manual testing

Testing notes:
- Regression test the current front pages without radio schedules - they should be unaffected by this change.
- Pages with radio schedules should now have the radio schedule rendered on the server, rather than appearing in after page load.
- If the fetch on the server for radio schedules data fails, the page should render without radio schedules. We should then attempt to fetch the data on the client and render the radio schedules component once if this data is successfully fetched.
- The integration test previously written for radio schedules can now be run, as we are rendering on the server.

### CODE REVIEW

File: src/app/routes/home/getInitialData/index.js [hunk 1]
```diff
@@ -14,13 +17,69 @@ const transformJson = pipe(
   filterGroupsWithoutStraplines,
 );
 
-export default async (path) => {
+export default async ({ path, service, variant = 'default' }) => {
+  // should we think of a resusable pattern for using config outside of our react app?
+  // such as const config = await getConfig(service);
```
Comment (tochwill) [inline src/app/routes/home/getInitialData/index.js]:
```suggestion
  // such as const config = await getConfig(service, variant);
```

File: src/server/index.jsx [hunk 2]
```diff
@@ -260,7 +260,11 @@ server
           routes,
           urlPath,
         );
-        const data = await route.getInitialData(url);
+        const data = await route.getInitialData({
```
Comment (tochwill) [inline src/server/index.jsx:263]: This has been updated across the board in https://github.com/bbc/simorgh/pull/6109 - we can pull those changes into this branch (or pull out into a smaller separate PR) if we want to merge the radio schedules work first.

**Comment by tochwill** [comment]
Dependent on https://github.com/bbc/simorgh/pull/6171/files

File: src/server/index.jsx (hunk 2 above)
Comment (tochwill) [reply src/server/index.jsx:263]: These changes are now in https://github.com/bbc/simorgh/pull/6171/files

File: src/app/lib/config/services/getConfig/index.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import pathOr from 'ramda/src/pathOr';
+
+/* This function allows for a dynamic import of service config outside of
+the react application (for example, for server side data fetching). Within
+the react application, the config can be found in the ServiceContext. */
+const getConfig = async (service, variant = 'default') => {
```
Comment (tochwill) [inline src/app/lib/config/services/getConfig/index.js]: I'm  a little uncertain about this - but this seems like the best way of being able to use the service config outside of the react application, which appears to be impossible at the moment? Using a normal static import of all of the config and pulling out the one we need would bloat the bundle massively. This means that we're importing the config for this service here in one way, and using a 'loadable' react context in the react application. This seems like a sensible option without a big refactor in the way be load config, but I'm very open to suggestions.

File: src/app/routes/home/getInitialData/index.js [hunk 4]
```diff
@@ -14,8 +17,26 @@ const transformJson = pipe(
...
+    ['radioSchedule', 'hasRadioSchedule'],
+    config,
+  );
+
+  const radioScheduleOnFrontPage = pathOr(
+    false,
+    ['radioSchedule', 'onFrontPage'],
+    config,
+  );
+
```
Comment (tochwill) [inline src/app/routes/home/getInitialData/index.js:34]: One thing we are not doing here is respecting the value of the toggle for radio schedules for this particular environment. We could import the static toggles json file and check that, but we are moving towards an approach where toggles come from an external endpoint, and are configured within the react application. So as is, if radio schedules is toggled off for an environment, we would still fetch the data for them on the server (and not render the component due to the toggles checks within the react application).

File: src/app/routes/utils/fetchPageData/index.js [hunk 5]
```diff
@@ -67,12 +67,11 @@ const handleError = (e) => {
   };
 };
 
-const fetchData = (pathname) => {
-  const url = getUrl(pathname);
-
+export const fetchData = (url) => {
```
Comment (tochwill) [inline src/app/routes/utils/fetchPageData/index.js]: We wanted to reuse all of the data fetching logic, but pass in an endpoint that was unrelated to the logic handled by `getUrl`. We should rearrange the tests for fetchPageData to test the fetchData function if we stick with this.

File: src/app/routes/utils/withRadioSchedule/index.test.js [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        failedPageDataPromise,
+        'hausa',
+        'http://localhost/mock-frontpage-path',
+      );
+
+      expect(json).toBeUndefined();
+      expect(status).toBe(404);
+    });
+  });
+});
```
Comment (tochwill) [inline src/app/routes/utils/withRadioSchedule/index.test.js:99]: We have tested so far:

- Successful data fetches (so the promise resolves)
- Unsuccessful data fetches (i.e. - the promise has resolved but the status code is not 2xx or 3xx)) 

How about if we encounter an actual error when fetching radio schedules data, so the promise is rejected? I think `fetchData` with currently throw an error that isn't caught anywhere which will give us a 500. If we want to render the page without radio schedules if the page data is fetched and radio schedules fails, we should handle this.

File: src/app/routes/utils/withRadioSchedule/index.js [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+  const { SIMORGH_APP_ENV, SIMORGH_BASE_URL } = process.env;
+
+  const radioScheduleUrl = getRadioScheduleEndpoint({
+    service,
+    env: SIMORGH_APP_ENV,
+    queryString: getQueryString(path),
+    baseUrl: SIMORGH_BASE_URL,
+  });
+
+  const radioSchedulePromise = fetchData(radioScheduleUrl);
```
Comment (tochwill) [inline src/app/routes/utils/withRadioSchedule/index.js:16]: If we want to still render the page if this promise _rejects_ we can wrap this in error handling that resolves with no data, which will allow Promise.all to resolve if the page data is successfully fetched.

File: src/integration/pages/articlePage/__snapshots__/mundo.test.js.snap [hunk 8]
```diff
@@ -1985,7 +1985,7 @@ exports[`Given I am on a Mundo Article page When I view the source code in the b
             crossorigin="anonymous"
             defer
     >
-      [21]
+      [19]
```
Comment (tochwill) [inline src/integration/pages/articlePage/__snapshots__/mundo.test.js.snap:1996]: I'm not sure why these numbers are being updated in these snapshots for the integration tests. Is it something to do with webpack chunking differently between test runs?

File: src/app/routes/utils/withRadioSchedule/index.test.js (hunk 6 above)
Comment (tochwill) [reply src/app/routes/utils/withRadioSchedule/index.test.js:99]: I was wrong here - `fetchData` catches the error, so this will render the page without radio schedules if the promise is rejected. I'm adding a test case.

File: src/app/containers/App/App.test.jsx [hunk 9]
```diff
@@ -185,6 +185,7 @@ describe('App', () => {
           expect(route.getInitialData).toHaveBeenCalledWith({
             path: pathname,
             service: 'news',
```
Comment (mmcculloch) [inline src/app/containers/App/App.test.jsx:187]: Should we change this service to one that has an actual variant?

File: src/app/containers/RadioSchedule/Canonical/index.jsx [hunk 10]
```diff
@@ -104,6 +104,9 @@ const CanonicalRadioSchedule = ({ endpoint }) => {
   } = useContext(ServiceContext);
 
   const { timeOnServer } = useContext(RequestContext);
+
+  const [schedule, setRadioSchedule] = useState(initialData);
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/Canonical/index.jsx]: Rename one of these so they match. e.g. `radioSchedule, setRadioSchedule`

File: src/app/containers/RadioSchedule/Canonical/index.jsx [hunk 11]
```diff
@@ -120,24 +123,26 @@ const CanonicalRadioSchedule = ({ endpoint }) => {
...
+          service,
+          timeOnServer || timeOnClient,
+        );
+        setRadioSchedule(schedules);
+      };
+
+      const fetchRadioScheduleData = (pathname) =>
+        fetch(pathname, { mode: 'no-cors' })
+          .then(handleResponse)
+          .catch((e) => logger.error(`HTTP Error: "${e}"`));
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/Canonical/index.jsx]: Update this to the new error format https://github.com/bbc/simorgh/blob/latest/src/app/lib/README.md#logging

File: src/app/containers/RadioSchedule/Canonical/index.jsx [hunk 12]
```diff
@@ -120,24 +123,26 @@ const CanonicalRadioSchedule = ({ endpoint }) => {
...
-    const fetchRadioScheduleData = (pathname) =>
-      fetch(pathname, { mode: 'no-cors' })
-        .then(handleResponse)
-        .catch((e) => logger.error(`HTTP Error: "${e}"`));
-
-    fetchRadioScheduleData(endpoint);
-  }, [endpoint, locale, script, service, timeOnServer, timezone]);
+    if (!schedule) {
+      const handleResponse = async (response) => {
+        const radioScheduleData = await response.json();
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/Canonical/index.jsx:129]: Check `response.ok` is `true` first, otherwise throw an error

Comment (mmcculloch) [reply src/app/containers/RadioSchedule/Canonical/index.jsx:129]: We can also look into extra logging if status codes are not what we expect (200)

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx [hunk 13]
```diff
@@ -4,52 +4,134 @@ import { matchSnapshotAsync } from '@bbc/psammead-test-helpers';
...
 
 describe('Canonical RadioSchedule', () => {
   afterEach(() => {
     fetch.resetMocks();
   });
 
-  it('renders correctly for a service with a radio schedule and page frequency URL', async () => {
-    fetch.mockResponse(JSON.stringify(arabicRadioScheduleData));
+  describe('With initial data', () => {
+    it('renders correctly for a service with initial data', async () => {
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/Canonical/index.test.jsx]: Can just be `renders correctly for a service`

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx [hunk 14]
```diff
@@ -4,52 +4,134 @@ import { matchSnapshotAsync } from '@bbc/psammead-test-helpers';
...
+      );
+      const { container } = render(
+        <RadioSchedulesWithContext
+          service="arabic"
+          initialData={initialData}
+          radioScheduleToggle
+        />,
+      );
+
+      await wait(() => {
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/Canonical/index.test.jsx:58]: Consider replacing with `act`

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx [hunk 15]
```diff
@@ -4,52 +4,134 @@ import { matchSnapshotAsync } from '@bbc/psammead-test-helpers';
...
+      const { container } = render(
+        <RadioSchedulesWithContext service="arabic" />,
+      );
+      await wait(() => {
+        expect(container).toBeEmpty();
+      });
+    });
+
+    it('does not render when data contains no programs', async () => {
+      fetch.mockResponse(
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/Canonical/index.test.jsx]: Consider testing for fetch failures as well.

File: src/app/containers/RadioSchedule/index.jsx [hunk 16]
```diff
@@ -6,9 +6,13 @@ import { ServiceContext } from '#contexts/ServiceContext';
 import { RequestContext } from '#contexts/RequestContext';
 import useToggle from '#hooks/useToggle';
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/index.jsx:7]: Double check the toggle is tested.

File: src/app/containers/RadioSchedule/utilities/processRadioSchedule.js [hunk 17]
```diff
@@ -22,13 +22,17 @@ export const getLink = (state, program, service) => {
 };
 
 export default (radioScheduleData, service, currentTime) => {
+  if (!radioScheduleData) {
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/utilities/processRadioSchedule.js]: Consider renaming to `data` and cleaning up var names

File: src/app/containers/RadioSchedule/utilities/processRadioSchedule.js [hunk 18]
```diff
@@ -22,13 +22,17 @@ export const getLink = (state, program, service) => {
...
+    return null;
+  }
+
   // finding latest program, that may or may not still be live. this is because there isn't
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/utilities/processRadioSchedule.js:30]: This comment needs to be moved now we think

File: src/app/containers/RadioSchedule/utilities/processRadioSchedule.js [hunk 19]
```diff
@@ -22,13 +22,17 @@ export const getLink = (state, program, service) => {
...
   // always a live program, in which case we show the most recently played program on demand.
+  const radioSchedules = pathOr([], ['schedules'], radioScheduleData);
```
Comment (mmcculloch) [inline src/app/containers/RadioSchedule/utilities/processRadioSchedule.js]: `const radioSchedules = radioScheduleData.schedules || [];` Could also work here?

File: src/app/pages/FrontPage/index.test.jsx [hunk 20]
```diff
@@ -34,7 +34,10 @@ let pageData;
 beforeEach(async () => {
   fetch.mockResponse(JSON.stringify(frontPageDataPidgin));
 
-  const response = await getInitialData('some-front-page-path');
+  const response = await getInitialData({
```
Comment (mmcculloch) [inline src/app/pages/FrontPage/index.test.jsx:37]: Check other occurrences of `getInitialData` to ensure an object is passed. Seems to not fail with the old string format.

File: src/app/routes/home/getInitialData/index.js [hunk 21]
```diff
@@ -14,8 +17,26 @@ const transformJson = pipe(
...
+  const pageHasRadioSchedule = hasRadioSchedule && radioScheduleOnFrontPage;
```
Comment (mmcculloch) [inline src/app/routes/home/getInitialData/index.js]: Move this into a separate function `pageHasRadioSchedule(service, variant)`

File: src/app/routes/home/getInitialData/index.test.js [hunk 22]
```diff
@@ -1,15 +1,92 @@
...
+    fetchMock.restore();
+  });
 
-  expect(pageData.metadata.language).toEqual('pcm');
-  expect(pageData.metadata.summary).toEqual(
-    'We dey give una latest tori on top politics, environment, business, sports, entertainment, health, fashion and all di oda things wey dey happen for West and Central Africa come add di rest of di world join. For better informate plus explanation of all di ogbonge tori wey pipo never hear about for inside West and Central Africa, BBC Pidgin dey serve am with video, audio, maps and oda graphics join.',
-  );
-  expect(pageData.promo.name).toEqual('Domot');
-  expect(pageData.content.groups.length).toBeTruthy();
+  it('should return essential data for a page without radio schedules to render', async () => {
```
Comment (mmcculloch) [inline src/app/routes/home/getInitialData/index.test.js]: Maybe reword to "data" rather than "essential data"

File: src/app/routes/home/getInitialData/index.test.js [hunk 23]
```diff
@@ -1,15 +1,92 @@
...
+      'Ziyarci shafin BBC Hausa domin samun rahotannin bidiyo da hotuna kan labarun Najeriya da Nijar da ma sauran sassan duniya baki daya.',
+    );
+    expect(pageData.promo.name).toEqual('Labaran Duniya');
+    expect(pageData.content.groups.length).toBeTruthy();
+
+    expect(pageData.radioScheduleData.length).toBe(4);
+  });
+
+  it('should return data for service with radio schedules, but without radio schedules on front page', async () => {
+    getConfig.mockImplementationOnce(() => ({
```
Comment (mmcculloch) [inline src/app/routes/home/getInitialData/index.test.js:66]: A missing combination that could be tested
```
    getConfig.mockImplementationOnce(() => ({
      radioSchedule: {
        hasRadioSchedule: false,
        onFrontPage: true,
      },
    }));
```

File: src/app/routes/utils/fetchPageData/index.js [hunk 24]
```diff
@@ -67,12 +67,11 @@ const handleError = (e) => {
...
   logger.info(DATA_REQUEST_RECEIVED, { url });
```
Comment (mmcculloch) [inline src/app/routes/utils/fetchPageData/index.js]: Can we add another error constant?

File: src/app/routes/utils/withRadioSchedule/index.test.js [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+        'hausa',
+        'http://localhost/mock-frontpage-path',
+      );
+
+      expect(json).toBeUndefined();
+      expect(status).toBe(404);
+    });
+
+    describe('fetch API promises rejected', () => {
+      it('should return page data without radio schedules if radio schedule fetch promise is rejected', async () => {
```
Comment (mmcculloch) [inline src/app/routes/utils/withRadioSchedule/index.test.js:115]: Would be good to test that this failure is being logged

File: src/app/routes/utils/withRadioSchedule/index.test.js [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+      });
+
+      const { json, status } = await withRadioSchedule(
+        failedPageDataPromise,
+        'hausa',
+        'http://localhost/mock-frontpage-path',
+      );
+
+      expect(json).toBeUndefined();
+      expect(status).toBe(404);
```
Comment (mmcculloch) [inline src/app/routes/utils/withRadioSchedule/index.test.js:86]: Would be good to test that this failure is being logged

File: src/app/routes/utils/withRadioSchedule/index.test.js [hunk 27]
```diff
@@ -0,0 +1,120 @@
...
```
Comment (mmcculloch) [inline src/app/routes/utils/withRadioSchedule/index.test.js:106]: Would be good to test that this failure is being logged

File: src/app/routes/utils/withRadioSchedule/index.js [hunk 28]
```diff
@@ -0,0 +1,35 @@
+import { getRadioScheduleEndpoint } from '#lib/utilities/getRadioSchedulesUrls';
+import { fetchData } from '../fetchPageData';
+import { getQueryString } from '#lib/utilities/urlParser';
+import processRadioSchedule from '#containers/RadioSchedule/utilities/processRadioSchedule';
+
+const withRadioSchedules = async (pageDataPromise, service, path) => {
```
Comment (mmcculloch) [inline src/app/routes/utils/withRadioSchedule/index.js]: Should be `withRadioSchedule` to match the file name

File: src/app/lib/config/services/getConfig/index.js [hunk 29]
```diff
@@ -0,0 +1,22 @@
...
+  let serviceConfig;
+
+  try {
+    const { service: config } = await import(`#lib/config/services/${service}`);
+    serviceConfig = config;
+  } catch (error) {
+    // eslint-disable-next-line no-console
+    console.error(
```
Comment (tochwill) [inline src/app/lib/config/services/getConfig/index.js]: Replace with the logger if we go with this approach.

File: src/app/routes/utils/withRadioSchedule/index.test.js [hunk 30]
```diff
@@ -0,0 +1,140 @@
+import fetchMock from 'fetch-mock';
+import loggerMock from '#testHelpers/loggerMock'; // Must be imported before fetchPageData
```
Comment (tochwill) [inline src/app/routes/utils/withRadioSchedule/index.test.js]:
```suggestion
import loggerMock from '#testHelpers/loggerMock';
```

File: src/app/routes/utils/fetchPageData/index.js (hunk 5 above)
Comment (tochwill) [reply src/app/routes/utils/fetchPageData/index.js]: Taking another look at this, if we want to handle errors differently, which we probably do, then we don't want to reuse this function. Have added the data fetching code as a function into the withRadioSchedules file which means we can handle errors with a different constant, and ignore all of the 'error propagation' that we don't need for this call - will push up changes after a conversation tomorrow.

**Comment by tochwill** [comment]
I've gone through and implemented changes for all of the points raised in the swarm review (and updated a couple more things that I noticed going back through the PR). I have _not_ yet made changes for the following, which could be handled today or in follow up tickets.

src/app/containers/App/App.test.jsx
Should we change this service to one that has an actual variant?

src/app/containers/RadioSchedule/Canonical/index.jsx
We can also look into extra logging if status codes are not what we expect (200)

src/app/pages/FrontPage/index.test.jsx
Check other occurrences of getInitialData to ensure an object is passed. Seems to not fail with the old string format.

src/app/routes/utils/fetchPageData/index.js
Can we add another error constant? - note: I have stopped trying to reuse this function, and instead have included error handling in the `withRadioSchedule` function. We can add a new constant and replace the usage of `DATA_FETCH_ERROR`.

src/app/containers/RadioSchedule/Canonical/index.test.jsx
Consider replacing with act

This is covered by this ticket: https://github.com/bbc/simorgh/issues/6219

File: src/app/lib/config/services/getConfig/index.js (hunk 3 above)
Comment (tochwill) [reply src/app/lib/config/services/getConfig/index.js]: What if we move this so it's not co-located with the config files, but only in the routes, where we need to access config like this?

**Comment by oluoluoxenfree** [comment]
@tochwill Could you fill in more of the PR description? :)

File: src/app/pages/FrontPage/index.test.jsx (hunk 20 above)
Comment (thekp) [reply src/app/pages/FrontPage/index.test.jsx:37]: > Seems to not fail with the old string format.

I feel like this test still passes because of what `fetch.mockResponse` is doing. It would return a response regardless of what the fetch path is and in the case where the path is not defined in the object, I assume it's `undefined` but the response is returned anyways. 

This should be updated in this issue: https://github.com/bbc/simorgh/issues/5960 to be URL sensitive when fetch mocking.

File: src/app/routes/home/getInitialData/index.js [hunk 31]
```diff
@@ -14,8 +17,30 @@ const transformJson = pipe(
...
+  );
+
+  return serviceHasRadioSchedule && radioScheduleOnFrontPage;
+};
+
+export default async ({ path, service, variant }) => {
+  const pageHasRadioSchedule = await hasRadioSchedule(service, variant);
+
+  const { json, ...rest } = pageHasRadioSchedule
+    ? await withRadioSchedule(fetchPageData(path), service, path)
```
Comment (thekp) [inline src/app/routes/home/getInitialData/index.js]: Is it worth changing this to accept an object?

**Review by thekp — COMMENTED** [review]
Looks good to me overall. Snapshots seem to be failing integration tests 👀  and a small comment about the `withRadioSchedule` function.

**Comment by tochwill** [comment]
Fixing merge conflicts

File: src/app/containers/RadioSchedule/Canonical/index.jsx [hunk 32]
```diff
@@ -91,55 +93,76 @@ const RadioFrequencyLink = styled.a`
...
+      };
+
+      const fetchRadioScheduleData = pathname =>
+        fetch(pathname, { mode: 'no-cors' })
+          .then(handleResponse(pathname))
+          .catch(error => {
+            logger.error(
+              JSON.stringify(
+                {
+                  event: DATA_FETCH_ERROR,
```
Comment (chris-hinds) [inline src/app/containers/RadioSchedule/Canonical/index.jsx]: would it be better to have a custom event for radio schedule data? it would make it easier to filter through the logs?

File: src/app/containers/RadioSchedule/Canonical/index.jsx [hunk 33]
```diff
@@ -91,55 +93,76 @@ const RadioFrequencyLink = styled.a`
...
   } = useContext(ServiceContext);
 
   const { timeOnServer } = useContext(RequestContext);
-  const header = pathOr(null, ['header'], radioSchedule);
+
+  const [radioSchedule, setRadioSchedule] = useState(initialData);
+
+  const header = pathOr(null, ['header'], radioScheduleConfig);
   const frequenciesPageUrl = pathOr(
     null,
```
Comment (ryanmccombe) [inline src/app/containers/RadioSchedule/Canonical/index.jsx]: There is a simpler `path` function for when we don't need to specify a fallback value: https://ramdajs.com/docs/#path

`path` falls back to undefined - I don't think `null` vs `undefined` makes a difference in most instances here

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx [hunk 34]
```diff
@@ -1,55 +1,173 @@
...
+      const initialData = processRadioSchedule(
+        arabicRadioScheduleData,
+        'arabic',
+        Date.now(),
+      );
+      await matchSnapshotAsync(
+        <RadioScheduleWithContext
+          service="arabic"
+          initialData={initialData}
+          radioScheduleToggle
```
Comment (itimirichard) [inline src/app/containers/RadioSchedule/Canonical/index.test.jsx]: Do we need the `service` and `radioScheduleToggle` props here and in the other instances of `RadioScheduleWithContext` below? They don't seem to be needed within the component on line 16 above

File: src/app/containers/RadioSchedule/Canonical/index.jsx (hunk 33 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.jsx]: I've just been looking at this too - I'll update the usage across this PR.

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx (hunk 34 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.test.jsx]: We don't! Will remove them.

File: src/app/containers/RadioSchedule/Canonical/index.jsx (hunk 32 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.jsx]: Sure - I added that to do in another issue, but I can make the change as part of this PR instead.

File: src/app/containers/RadioSchedule/Canonical/index.jsx (hunk 33 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.jsx]: Actually as all of these examples are just one level deep into the object, I should be fine to default `radioScheduleConfig` to an empty object, and then destructure these values out.

File: src/app/routes/home/getInitialData/index.js (hunk 31 above)
Comment (tochwill) [reply src/app/routes/home/getInitialData/index.js]: Done!

File: src/app/containers/RadioSchedule/Canonical/index.jsx (hunk 33 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.jsx]: Done! Have updated usages across this PR.

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx (hunk 34 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.test.jsx]: Done!

File: src/app/containers/RadioSchedule/Canonical/index.jsx (hunk 32 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.jsx]: Done! I will add to sumologic when I've made all requested changes and merged.

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx [hunk 35]
```diff
@@ -1,55 +1,153 @@
...
+      );
+      const { container } = render(
+        <RadioScheduleWithContext initialData={initialData} />,
+      );
 
-    await matchSnapshotAsync(
-      <RadioSchedulesWithContext service="arabic" radioScheduleToggle />,
-    );
-  });
+      await wait(() => {
```
Comment (DenisHdz) [inline src/app/containers/RadioSchedule/Canonical/index.test.jsx:58]: Looks good to me overall 💯
Wondering if we still have to replace this with `await act( async () => {`

**Review by itimirichard — APPROVED** [review]
Thanks for making all these changes Tom. Looks good to me!

File: src/app/containers/RadioSchedule/Canonical/index.test.jsx (hunk 35 above)
Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.test.jsx:58]: I would like to, I've added that to another ticket to do once this is reviewed, there's quite a lot of changes in here already!

Comment (tochwill) [reply src/app/containers/RadioSchedule/Canonical/index.test.jsx:58]: https://github.com/bbc/simorgh/issues/6219

Comment (DenisHdz) [reply src/app/containers/RadioSchedule/Canonical/index.test.jsx:58]: Excellent 👍

**Review by DenisHdz — APPROVED** [review]

**Comment by paruchurisilpa** [comment]
* Regression testing on Radio schedules didnt show any issues.
   *  If the fetch on the server for radio schedules data fails, the page should render without radio schedules. We should then attempt to fetch the data on the client and render the radio schedules component once if this data is successfully fetched  --> This I have checked with help from Tom.
   * Integration tests have passed.

LGTM...

**Comment by tochwill** [comment]
Thank you @paruchurisilpa :) I just noticed that the integration test is still commented out, I have amended it slightly and will push up to this branch before merging this in.
